apis/metal3.io/v1alpha1/hostfirmwaresettings_types.go

```diff
@@ -0,0 +1,169 @@
+/*
+
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	"strconv"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// NOTE: json tags are required.  Any new fields you add must have json tags for the fields to be serialized.
+
+// WriteableFirmwareSetting describes a FirmwareSetting whose value can be changed via Ironic clean-steps.
+type WriteableFirmwareSetting struct {
+	// Identifier for the Firmware setting.
+	Name string `json:"name" required:"true"`
+
+	// Value of the Firmware setting.
+	Value string `json:"value" required:"true"`
+}
+
+// FirmwareSetting describes fields read from Ironic for a particular setting in the BMC.
+type FirmwareSetting struct {
+	// Identifier for the Firmware setting.
+	Name string `json:"name" required:"true"`
+
+	// Value of the Firmware setting.
+	Value string `json:"value" required:"true"`
+
+	// The type of setting.
+	// +kubebuilder:validation:Enum=Enumeration;String;Integer;Boolean;Password
+	AttributeType string `json:"attribute_type,omitempty"`
+
+	// The allowable value for an Enumeration type setting.
+	AllowableValues []string `json:"allowable_values,omitempty"`
+
+	// The lowest value for an Integer type setting.
+	LowerBound *int `json:"lower_bound,omitempty"`
+
+	// The highest value for an Integer type setting.
+	UpperBound *int `json:"upper_bound,omitempty"`
+
+	// Minimum length for a String type setting.
+	MinLength *int `json:"min_length,omitempty"`
+
+	// Maximum length for a String type setting.
+	MaxLength *int `json:"max_length,omitempty"`
+
+	// Whether or not this setting is read only.
+	ReadOnly *bool `json:"read_only,omitempty"`
+
+	// Whether or not a reset is required after changing this setting.
+	ResetRequired *bool `json:"reset_required,omitempty"`
+
+	// Whether or not this setting's value is unique to this node, e.g.
+	// a serial number.
+	Unique *bool `json:"unique,omitempty"`
+}
+
+// HostFirmwareSettingsSpec defines the desired state of HostFirmwareSettings
+type HostFirmwareSettingsSpec struct {
+	// Important: Run "make" to regenerate code after modifying this file
+
+	WriteableFirmwareSettings []WriteableFirmwareSetting `json:"writeableFirmwareSettings,omitempty"`
```

Comment:
Is there a reason we are not just making this a `map[string]string`? I tend to prefer that as it eliminates the possibility of duplicate values and therefore the question of what to do about them. (On the other hand I have used it on occasion in APIs and it turned out later to be the Wrong Thing.)

If the reason is related to problems generating an open-ended OpenAPI schema then I believe there are workarounds available.

Reply:
We should probably add the directives:

```
// + patchMergeKey=name
// +patchStrategy=merge
```

so that people can easily just patch in only the values they care about.

(Merge key is only needed if we keep the list-of-structs type.)

Reply:
I like using the map as it will simplify things when the spec is updated and we need to validate the value. But yeah I ran into some issues when trying to generate the CRD. I will look into using this more.

Reply:
Updated to use maps for status and spec and for the firmwareschema. The generated yaml seems fine using "additionalProperties".